### PR TITLE
Move CI

### PR DESCRIPTION
**Pull request by MaksimZhukov**

### CODE REVIEW

File: .github/workflows/build-node-packages.yml [hunk 1]
```diff
@@ -0,0 +1,202 @@
+name: Generate Node.js
```
Comment (maxim-lobanov) [inline .github/workflows/build-node-packages.yml]: `Generate Node.js package`

File: .github/workflows/build-node-packages.yml [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+        required: true
+        default: 'false'
+
+env:
+  VERSION: ${{ github.event.inputs.VERSION }}
+  ARCHITECTURE: x64
+
+jobs:
+  build_node:
+    name: Build Node ${{ github.event.inputs.VERSION }} ${{ matrix.platform }}
```
Comment (maxim-lobanov) [inline .github/workflows/build-node-packages.yml]: `Build Node.js ${{ github.event.inputs.VERSION }} [${{ matrix.platform }}]`

File: .github/workflows/build-node-packages.yml [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+        ./builders/build-node.ps1 -Version $env:VERSION `
+                                  -Platform ${{ matrix.platform }} `
+                                  -Architecture $env:ARCHITECTURE
+      shell: pwsh
+
+    - name: Publish artifact
+      uses: actions/upload-artifact@v2
+      with:
+        name: node-${{ env.VERSION }}-${{ matrix.platform }}
+        path: /home/runner/work/node-versions/artifact
```
Comment (maxim-lobanov) [inline .github/workflows/build-node-packages.yml]: Let's move this dir under RUNNER_TEMP and get rid of hardcoded path

File: .github/workflows/build-node-packages.yml [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+      matrix:
+        platform: [linux, darwin, win32]
+    steps:
+    - uses: actions/checkout@v2
+      with:
+        submodules: true
+
+    - name: Create artifact directories
+      run: |
+        binariesDirectory=$RUNNER_WORKSPACE/binaries
```
Comment (maxim-lobanov) [inline .github/workflows/build-node-packages.yml]: Why we need to create this directories here and define new environment variables?
Let's do that in `build-node.ps1`

File: .github/workflows/build-node-packages.yml [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+        echo ::set-env name=BINARIES_DIRECTORY::$binariesDirectory
+        mkdir $binariesDirectory
+
+        artifactDirectory=$RUNNER_WORKSPACE/artifact
+        echo ::set-env name=ARTIFACT_DIRECTORY::$artifactDirectory
+        mkdir $artifactDirectory
+
+    - name: Build Node ${{ env.VERSION }}
```
Comment (maxim-lobanov) [inline .github/workflows/build-node-packages.yml]: Build Node.js ${{ env.VERSION }}

File: .github/workflows/build-node-packages.yml [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+
+  test_node:
+    name: Test Node ${{ github.event.inputs.VERSION }} ${{ matrix.platform }}
+    needs: build_node
+    runs-on: ${{ matrix.os }}-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+        - os: ubuntu
```
Comment (maxim-lobanov) [inline .github/workflows/build-node-packages.yml]: let's define os fully here like `ubuntu-latest`, `macos-latest`

File: .github/workflows/build-node-packages.yml [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+    steps:
+    - uses: actions/checkout@v2
+      with:
+        submodules: true
+
+    - name: Set AGENT_TOOLSDIRECTORY variable
+      if: matrix.platform == 'win32'
+      run: |
+        # GitHub Windows images don't have `AGENT_TOOLSDIRECTORY` variable
+        echo ::set-env name=AGENT_TOOLSDIRECTORY::$RUNNER_TOOL_CACHE
```
Comment (maxim-lobanov) [inline .github/workflows/build-node-packages.yml]: How setup-node works on GitHub Windows machines in this case?

File: .github/workflows/build-node-packages.yml [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+      shell: bash
+
+    - name: Fully cleanup the toolcache directory before testing
+      run: |
+        ./helpers/clean-toolcache.ps1 -ToolName "node"
```
Comment (maxim-lobanov) [inline .github/workflows/build-node-packages.yml]: `run: helpers/clean-toolcache.ps1 -ToolName "node"`

File: .github/workflows/build-node-packages.yml [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+      shell: pwsh
+
+    - name: Setup node ${{ env.VERSION }}
+      uses: actions/setup-node@v2.1.1
+      with:
+        node-version: ${{ env.VERSION }}
+
+    - name: Wait for the logs
+      run: |
+        Write-Host "Fake step that do nothing"
```
Comment (maxim-lobanov) [inline .github/workflows/build-node-packages.yml:98]: Do we need it on GH?

File: .github/workflows/build-node-packages.yml [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+      run: |
+        Install-Module Pester -Force -Scope CurrentUser -RequiredVersion 4.10.1
+        Import-Module Pester
+        $pesterParams = @{
+          Path="./Node.Tests.ps1";
+          Parameters=@{
+            Version="$env:VERSION";
+          }
+        }
+        Invoke-Pester -Script $pesterParams -EnableExit -OutputFile "test_results.xml" -OutputFormat NUnitXml
```
Comment (maxim-lobanov) [inline .github/workflows/build-node-packages.yml]: Why we pass `OutputFile` and `OutputFormat` if we don't use it.
Is it possible to switch to Pester 5 Btw?

File: .github/workflows/build-node-packages.yml [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+      working-directory: ./tests
+      shell: pwsh
+
+  publish_release:
+    name: Publish release
+    if: github.event.inputs.PUBLISH_RELEASES == 'true'
+    needs: test_node
+    runs-on: ubuntu-latest
+    outputs:
```
Comment (maxim-lobanov) [inline .github/workflows/build-node-packages.yml]: What is it?

File: .github/workflows/build-node-packages.yml [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+    - name: Publish Release ${{ env.VERSION }}
+      id: create_release
+      uses: actions/create-release@v1
+      env:
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+      with:
+        tag_name: ${{ env.VERSION }}-${{ github.run_id }}
+        release_name: ${{ env.VERSION }}
+        body: |
+          Upload Node.js ${{ env.VERSION }}
```
Comment (maxim-lobanov) [inline .github/workflows/build-node-packages.yml]: Why we need `Upload` word in body?

File: .github/workflows/build-node-packages.yml [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+
+  upload_assets:
+    name: Upload assets for ${{ matrix.platform }}
+    needs: publish_release
+    runs-on: ubuntu-latest
+    env:
+      UPLOAD_URL: ${{ needs.publish_release.outputs.upload_url }}
+    strategy:
+      matrix:
+        extension: ['tar.gz']
```
Comment (maxim-lobanov) [inline .github/workflows/build-node-packages.yml]: This matrix looks very strange, can we describe all cases in include?

File: .github/workflows/build-node-packages.yml [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+    - uses: actions/download-artifact@v2
+      with:
+        name: node-${{ env.VERSION }}-${{ matrix.platform }}
+
+    - name: Upload Release Asset
+      uses: actions/upload-release-asset@v1
+      env:
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+      with:
+        upload_url: ${{ env.UPLOAD_URL }}
```
Comment (maxim-lobanov) [inline .github/workflows/build-node-packages.yml]: Why can't we upload all artifacts in the same job? Or even in job where you publish release? 
Is `actions/upload-release-asset` or `actions/download-artifact` supports `*`?

File: .github/workflows/build-node-packages.yml [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+        asset_path: ./node-${{ env.VERSION }}-${{ matrix.platform }}-${{ env.ARCHITECTURE }}.${{ matrix.extension }}
+        asset_name: node-${{ env.VERSION }}-${{ matrix.platform }}-${{ env.ARCHITECTURE }}.${{ matrix.extension }}
+        asset_content_type: application/zip
+
+  create_pr:
```
Comment (maxim-lobanov) [inline .github/workflows/build-node-packages.yml]: trigger_pr

File: .github/workflows/build-node-packages.yml [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+    name: Create Pull Request
```
Comment (maxim-lobanov) [inline .github/workflows/build-node-packages.yml]: `Trigger Pull Request`

File: .github/workflows/build-node-packages.yml [hunk 17]
```diff
@@ -0,0 +1,202 @@
...
+        submodules: true
+
+    - name: Trigger "Create Pull Request" workflow
+      uses: actions/github-script@v2
+      with:
+        github-token: ${{ secrets.PERSONAL_TOKEN }}
+        script: |
+          # TODO: currently 'actions.createWorkflowDispatch' function does not work. I will investigate  
+          github.repos.createDispatchEvent({
+            owner: context.repo.owner,
```
Comment (maxim-lobanov) [inline .github/workflows/build-node-packages.yml:161]: `context.repo.owner` and `context.repo.repo` are available only inside github-script?

Comment (maxim-lobanov) [reply .github/workflows/build-node-packages.yml:161]: Yep, confirmed

File: .github/workflows/create-pr.yml [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+on: 
+# TODO: currently workflow dispatch endpoint does not work. I will investigate
+  repository_dispatch:
+    types: [create-pr]
+  workflow_dispatch:
+jobs:
+  build:
+    name: Create Pull Request
+    env:
+      REPOSITORY_NAME: 'node-versions'
```
Comment (maxim-lobanov) [inline .github/workflows/create-pr.yml]: Let's modify `manifest-generator.ps1` and avoid hardcoded repo name

File: tests/Node.Tests.ps1 [hunk 19]
```diff
@@ -6,11 +6,17 @@ param (
...
 
 function Get-UseNodeLogs {
-    $logsFolderPath = Join-Path -Path $env:AGENT_HOMEDIRECTORY -ChildPath "_diag" | Join-Path -ChildPath "pages"
+    $homeDir = $env:HOME
+    if ([string]::IsNullOrEmpty($homeDir)) {
+        # GitHub Windows images don't have `HOME` variable
+        $homeDir = $env:HOMEDRIVE
+    }
+
+    $logsFolderPath = Join-Path -Path $homeDir -ChildPath "runners/*/_diag/pages" -Resolve
```
Comment (maxim-lobanov) [inline tests/Node.Tests.ps1]: Is there any variable for agent folder?
`runners` can contain more than 1 agent

File: .github/workflows/build-node-packages.yml (hunk 1 above)
Comment (MaksimZhukov) [reply .github/workflows/build-node-packages.yml]: done

File: .github/workflows/build-node-packages.yml (hunk 2 above)
Comment (MaksimZhukov) [reply .github/workflows/build-node-packages.yml]: done

File: .github/workflows/build-node-packages.yml (hunk 3 above)
Comment (MaksimZhukov) [reply .github/workflows/build-node-packages.yml]: done

File: .github/workflows/build-node-packages.yml (hunk 4 above)
Comment (MaksimZhukov) [reply .github/workflows/build-node-packages.yml]: done

File: .github/workflows/build-node-packages.yml (hunk 5 above)
Comment (MaksimZhukov) [reply .github/workflows/build-node-packages.yml]: done

File: .github/workflows/build-node-packages.yml (hunk 6 above)
Comment (MaksimZhukov) [reply .github/workflows/build-node-packages.yml]: done

File: .github/workflows/build-node-packages.yml (hunk 7 above)
Comment (MaksimZhukov) [reply .github/workflows/build-node-packages.yml]: discussed

File: .github/workflows/build-node-packages.yml (hunk 8 above)
Comment (MaksimZhukov) [reply .github/workflows/build-node-packages.yml]: done

File: .github/workflows/build-node-packages.yml (hunk 9 above)
Comment (MaksimZhukov) [reply .github/workflows/build-node-packages.yml:98]: yes, discussed

File: .github/workflows/build-node-packages.yml (hunk 11 above)
Comment (MaksimZhukov) [reply .github/workflows/build-node-packages.yml]: where?)

File: .github/workflows/build-node-packages.yml (hunk 12 above)
Comment (MaksimZhukov) [reply .github/workflows/build-node-packages.yml]: removed

File: .github/workflows/build-node-packages.yml (hunk 13 above)
Comment (MaksimZhukov) [reply .github/workflows/build-node-packages.yml]: removed matrix

File: .github/workflows/build-node-packages.yml (hunk 14 above)
Comment (MaksimZhukov) [reply .github/workflows/build-node-packages.yml]: changed approach

File: .github/workflows/build-node-packages.yml (hunk 15 above)
Comment (MaksimZhukov) [reply .github/workflows/build-node-packages.yml]: done

File: .github/workflows/build-node-packages.yml (hunk 16 above)
Comment (MaksimZhukov) [reply .github/workflows/build-node-packages.yml]: done

File: .github/workflows/create-pr.yml (hunk 18 above)
Comment (MaksimZhukov) [reply .github/workflows/create-pr.yml]: done

File: .github/workflows/build-node-packages.yml (hunk 10 above)
Comment (MaksimZhukov) [reply .github/workflows/build-node-packages.yml]: I've removed options.
Need to update to Pester 5

File: tests/Node.Tests.ps1 (hunk 19 above)
Comment (MaksimZhukov) [reply tests/Node.Tests.ps1]: Currently we have two folders in the `runners` dir: `2.272.0` and `2.273.0`. Only `2.272.0` runner has the `pages` folder in the `_diag` dir.

File: .github/workflows/build-node-packages.yml (hunk 10 above)
Comment (MaksimZhukov) [reply .github/workflows/build-node-packages.yml]: Updated

**Review by maxim-lobanov — APPROVED** [review]

**Pull request merged**